### ISSUE TITLE
Filter support for ASP.NET Core metrics

### ISSUE THREAD

**Issue by vishweshbankwar** [open-telemetry/opentelemetry-dotnet#4982]
Some pre-release versions of Metrics instrumentation provided support for [filtering](https://github.com/open-telemetry/opentelemetry-dotnet/blob/1d3bb4b977387c3f48c45aca36a760b2a36d7c8a/src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs#L72) out measurements for `http.server.request.duration`. This option is removed in https://github.com/open-telemetry/opentelemetry-dotnet/pull/4981. Opening this issue to track any use cases that filtering may be needed for metrics.

**Comment by ZRich97** [comment]
Adding comment from discussion on https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1426:

> The main use case in my mind is filtering out unused metrics (noise). Some of our supported services don't need to collect metrics for every route in the application (and they currently don't in their non-OTel monitoring solutions). This filtering could also be done at the alerting/monitoring stage (metrics we don't care about could be identified by the dimensions collected/enriched by this instrumentation and filtered out of any dashboards/alerts), but why not avoid collecting unused metrics altogether?

Metric-level filters seem like the best way to reduce possible noise and collect metrics only for the routes/requests service owners are most interested in. Is there another OTel construct or tool aside from the Instrumentation packages that would fill the same role in its absence? Is there any downside to having filter options available? 

Our team is starting to move towards OTel across our .NET services. If there is a better approach we'd be interested in trying it out as well. Thanks!

**Comment by shatl** [comment]
Please bring filtering back. As it was mentioned above, one use case is to reduce noise caused by healthchecks

**Comment by akoken** [comment]
I really don't understand the reason behind removing already implemented features. Performance or allocation? Even if so, that's a tradeoff which should be considered by the consumer. IMHO, Enrich and Filter should be de facto for all instrumentations.

We operate many microservices, each with multiple instances, within the Kubernetes environment. As you may know, it's best practice to use liveness and readiness probes for health check in K8s. However, it's worth noting that they contribute to increased metric noise.

```bash
// Should be filtered
/health/live
/health/ready
/metrics
```

**Comment by joebeernink** [comment]
Adding my vote here for filtering out routes for health checks. 
1. Since metrics are created in the AppMetrics table in LogAnalytics, adding all of these unnecessary routes increases our Azure costs
2. Filtering these calls out on the query side of metrics means a lot of manual updating of queries when calculating overall Availability of the service for customer facing queries and for calculating service latencies. Since these calls are made very frequently, a service that gets relatively few calls from the customer could show wrong customer facing stats without the filters.

Basically, lack of this feature drives up cost of operations for the system, and the consumer should be able to choose which costs are bearable and which are not.

**Comment by WeihanLi** [comment]
Also hoping to see the feature, while think it may be better to support from the .NET API

also link to an issue in dotnet/aspnetcore https://github.com/dotnet/aspnetcore/issues/50654

**Comment by Elter71** [comment]
Any update on this issue?

**Comment by xkursatx** [comment]
by adding OpenTelemetry.Instrumentation.AspNetCore
and

```
builder.Services.AddOpenTelemetry()
    .WithMetrics(builder =>
    {
        builder
            .AddPrometheusExporter()
            .AddMeter("Microsoft.AspNetCore.Hosting", "Microsoft.AspNetCore.Server.Kestrel");

    })
    .WithTracing(builder =>
    {
        builder
            .AddAspNetCoreInstrumentation((options) => options.Filter = httpContext =>
            {
                if (httpContext.Request.Path.Value == "/metrics")
                {
                    return false;
                }
                return true;
            });
    });
```

i think this will resolve my problem. not tested yet but seems bright enough

**Comment by vishweshbankwar** [comment]
Thanks everyone for the feedback - The feature is requested for .NET9.0. Please track https://github.com/dotnet/aspnetcore/issues/50654 for updates.
